kuma/wiki/management/commands/render_document.py

```diff
@@ -2,100 +2,115 @@
-
-    def do_render(self, doc):
-        # Skip very recently rendered documents. This should help make it
-        # easier to start and stop an --all command without needing to start
-        # from the top of the list every time.
-        if doc.last_rendered_at:
-            now = datetime.datetime.now()
-            render_age = now - doc.last_rendered_at
-            min_age = datetime.timedelta(seconds=self.options['min_age'])
-            if (render_age < min_age):
-                logging.debug(u"Skipping %s (%s) - rendered %s sec ago" %
-                              (doc, doc.get_absolute_url(), render_age))
-                return
-
-        if self.options['force']:
-            doc.render_started_at = None
+                doc = Document.objects.get(locale=locale, slug=slug)
+                if self.options['force']:
+                    doc.render_started_at = None
+                log.info(u'Rendering %s (%s)' % (doc, doc.get_absolute_url()))
+                try:
+                    render_document(doc, self.cache_control, self.base_url)
+                    log.debug(u'DONE.')
+                except DocumentRenderingInProgress:
+                    log.error(
+                        u'Rendering is already in progress for this document.')
 
-        if self.options['nocache']:
-            cc = 'no-cache'
-        else:
-            cc = 'max-age=0'
+    def chain_render_docs(self, docs):
+        tasks = []
+        count = 0
+        total = docs.count()
+        mod = int(ceil(total * 0.2))
 
-        if self.options['defer']:
-            logging.info(u"Queuing deferred render for %s (%s)" %
-                          (doc, doc.get_absolute_url()))
-            render_document.delay(doc.pk, cc, self.base_url)
-            logging.debug(u"Queued.")
+        for doc in docs.iterator():
+            count += 1
 
-        else:
-            logging.info(u"Rendering %s (%s)" %
-                         (doc, doc.get_absolute_url()))
-            try:
-                render_document(doc, cc, self.base_url)
-                logging.debug(u"DONE.")
-            except DocumentRenderingInProgress:
-                logging.error(u"Rendering is already in progress for this document")
+            if self.options['force']:
+                doc.render_started_at = None
+
+            tasks.append(render_document.si(doc.pk, self.cache_control,
+                                            self.base_url))
+            if count % mod == 0 and count != total:
+                percent_complete = int(ceil((count / total) * 100))
+                tasks.append(
+                    email_render_document_progress.si(percent_complete,
+                                                      total))
+
+        # Add a 100% complete email task.
+        tasks.append(email_render_document_progress.si(100, total))
+
+        # Make it so.
+        chain(*tasks).apply_async()
```

Comment:
Inside the `document_render` task I can wrap the actual render call in a `try/except` block. But, I'd need to know what exceptions we want to retry the task and what exceptions we want to log or email?